GoogleForCreators/web-stories-wp
# Right Click Menu: Add copy/paste/clear style functionality

## Context

[Right Click Context Menu Designs](https://www.figma.com/file/1YM3jA9h4QGc4xLdyaXIJS/Stories-Sprint-1.7?node-id=7347%3A60832)

## Summary

Adds functionality for copying, pasting, and clearing styles (resetting to default) from elements. Next sprint we will add this functionality to text and shape elements too.

Note: it already works for all element types. Once the menus are built out we can drop them in and they should work.

## Relevant Technical Choices

Pasting styles can only work between elements of the *same type*.

*Shape*: the `square` shape is the only shape that can have a border and a border radius. Keeping `border` and `borderRadius` as non-clearable attributes so that pasting will not update those styles. 

## To-do

These actions are not in the correct place. Once #8382 is merged, they will be moved to the foreground media element menu

## User-facing changes

<details>
<summary>Gifs showing new functionality</summary>

|Element|Copy/Paste Styles|Clear Styles|
|--|--|--|
|`image`|![image](https://user-images.githubusercontent.com/22185279/127410013-321e1bfa-72da-4657-8aa3-90435f9e3e9e.gif)|![clear-image](https://user-images.githubusercontent.com/22185279/127409661-911ced63-db14-4de0-bad9-a48309dc8f82.gif)|
|`video`|![video](https://user-images.githubusercontent.com/22185279/127409681-3739b65c-aafc-4f67-b5ed-bc97e8a9d5a1.gif)|![clear-video](https://user-images.githubusercontent.com/22185279/127409683-3b909ae7-aee2-4e65-837b-8f6c693d95fc.gif)|
|`gif`|![gif](https://user-images.githubusercontent.com/22185279/127409696-b2d6c49a-7476-497f-b84b-11bbf1f63054.gif)|![clear-gif](https://user-images.githubusercontent.com/22185279/127409699-332faf95-e1f2-485e-891a-e09c92cbda61.gif)|
|`text`|![text](https://user-images.githubusercontent.com/22185279/127409712-cfaf8e2e-9f91-48ae-b714-c19f7f4c4809.gif)|![clear-text](https://user-images.githubusercontent.com/22185279/127409715-db3eefab-4bd5-4f13-b2ce-2870522770a6.gif)|
|`shape`|![shape](https://user-images.githubusercontent.com/22185279/127410629-5373bb4d-e289-4f5e-afd8-507bf3fbbad5.gif)|![shape-clear](https://user-images.githubusercontent.com/22185279/127410634-46e0654a-d16c-475a-99f0-a13fb6d6b75a.gif)|
</details>

## Testing Instructions

Testing this PR will be tedious because you'll have to test each type of element. Those types are
- `image`
- `video`
- `gif`
- `shape`
- `text`

For each element type:
1. Add two different elements to the canvas
2. Update some styles in the inspector panel fo rthat element
3. Right click that element and select `Copy Image Styles` (text to be updated in the future)
4. Right click the other element of the same type. Select `Paste image styles`
5. Should see updated styles on the element and in the inspector
6. Right click either element
7. Select `Clear image styles`
8. Should see styles in table reverted to default values

*Note*: all text types (title 1, title 2, paragraph, etc.) will have the _same_ defaults because they are all different stylings of the same *text* element.

The properties that you will see changed for pasting and clearing are outlined in this table:

|Element type|Properties that are updated when pasting or clearing styles|
|--|--|
|`image`, `video`, and `gif`|-`border`<br/>- `borderRadius`<br/>- `overlay`<br/>- `opacity`<br/>-`flip`<br/>- `rotationAngle`<br/>- `lockAspectRatio`|
|`shape`|- `backgroundColor`<br/>- `opacity`<br/>- `flip`<br/>- `rotationAngle`<br/>- `lockAspectRatio`|
|`text`|- `backgroundColor`<br/>- `backgroundTextMode`<br/>- `border`<br/>- `borderRadius`<br/>- `padding`<br/>- `textAlign`<br/>- `opacity`<br/>-`flip`<br/>- `rotationAngle`<br/>- `lockAspectRatio`|

### QA






## Reviews

### Does this PR have a security-related impact?

no

### Does this PR change what data or activity we track or use?

no

### Does this PR have a legal-related impact?

no

## Checklist




---



Fixes #7871
